File: day1/src/bin/part2.rs
use aoc::*;

fn main() {
    let res: usize = parser::chars::<char>()
        .map(|c| match c {
            '(' => 1,
            ')' => -1,
            _ => 0,
        })
        .scan(0, |acc, el| {
            *acc += el;
            Some(*acc)
        })
        .position(|level| level == -1)
        .unwrap();

   aoc::answer!(
        "The position of the character that causes Santa to first enter the basement is: {}.",
        res
    );
}

File: day1/src/bin/part2.ml
open Scanf

exception Error of string

let file = Sys.argv.(1)
let scanner = Scanning.from_file file

let rec _compute scanner current pos =
        try
                match bscanf scanner "%c" (fun x -> x) with
                | _ when current = -1 -> pos
                | '(' -> _compute scanner (current + 1) (pos + 1)
                | ')' -> _compute scanner (current - 1) (pos + 1)
                | '\n' | ' ' -> raise (Error "Never went to the floor -1")
                | c -> raise (Error (Printf.sprintf "Unknown character %c" c))
        with 
        End_of_file -> pos

let compute scanner =
        _compute scanner 0 0

let res = compute scanner;;
Printf.printf "The position of the instruction that brings santa to enter the basement is %d\n" res;;

